src/library_async.js

```diff
@@ -747,8 +808,9 @@ mergeInto(LibraryManager.library, {
 #endif
         Asyncify.state = Asyncify.State.Normal;
         runAndAbortIfError(Module['_asyncify_stop_rewind']);
-        Asyncify.freeData(Asyncify.currData);
+        _free(Asyncify.currData);
         Asyncify.currData = null;
+        noExitRuntime = false;
```

Comment:
This change is surprising, why is it here?

Reply:
This is also something I forgot to remove from `handleSleep` when I made fibers no longer use that function, but I suspect that there's actually a bug here. Initially this was added because the runtime would never shut down once a fiber switch has occurred, even after going back to the main context. But this actually seems to be the case for *any* use of `handleSleep`: it sets `noExitRuntime` to `true` at the very beginning but never restores it, as far as I can tell. I have not tested this thoroughly, though.

Reply:
Oh, it is normal for `handleSleep` to set that. Basically, anything that starts and async operation, which will let the program appear to finish but it is expected to resume later, must not exit the runtime. So `noExitRuntime` must be set to false in the Fibers code if it doesn't use `handleSleep`.

(It is true that in principle the code should restore the old value once any async operation is complete, but emscripten hasn't tried to track that carefully; once you go async, we assume the runtime won't be shut down. As the default is to not shut it down anyhow, this hasn't been an issue in practice.)

Reply:
Yes I'm aware, the fibers code sets it to `true` [here](https://github.com/emscripten-core/emscripten/pull/9859/files#diff-e1dcd8e7fa0139253f677464734f4f4cR988), and restores it back to `false` [here](https://github.com/emscripten-core/emscripten/pull/9859/files#diff-e1dcd8e7fa0139253f677464734f4f4cR924) just before rewinding or entering a new context. That seems to work well so far. The test deliberately uses `EXIT_RUNTIME=1` and C++ destructors to make sure that works.

As far as I can tell, `noExitRuntime` only needs to be `true` while unwinding, and when that's done it's safe to just set it to `false`. Is that correct or is there more to it? The `noExitRuntime = false` that I put in `handleSleep` happens after rewinding, just before the execution resumes. I believe that should be safe. But it makes sense to revert this change as not Fibers related. Let me know if I should do that.

Reply:
That sounds correct.

Reply:
However, it's not ok to change noExitRuntime to false here. It should change to the previous value, which may have been true anyhow.

I think it's fine to set it to true when doing an async operation, and never set it again - as I said earlier, that's imperfect, but it's what the rest of the code does, and it's good enough. We can consider fixing that separately from this PR.

Reply:
Alright. I'm definitely not a fan of this behavior, but I do agree it's better fixed outside of this PR. At the very least, it should be clearly documented. I only found out about it experimentally, and thought I genuinely found a bug.